File: market-data/crypto-news/examples/javascript/social.js
/**
 * Free Crypto News API - Social Media Examples
 * https://github.com/nirholas/free-crypto-news
 * 
 * Social media feeds, sentiment, and community tracking.
 */

const BASE_URL = 'https://cryptocurrency.cv';

// =============================================================================
// X (TWITTER) ENDPOINTS
// =============================================================================

/**
 * Get X/Twitter feed for crypto topics
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Twitter feed data
 */
async function getXFeed(options = {}) {
  const params = new URLSearchParams({
    topic: options.topic || 'crypto',
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/x?${params}`);
  return response.json();
}

/**
 * Get X/Twitter mentions for a specific asset
 * @param {string} asset - Asset symbol (BTC, ETH, etc.)
 * @returns {Promise<Object>} Mentions data
 */
async function getXMentions(asset) {
  const response = await fetch(`${BASE_URL}/api/social/x/mentions?asset=${asset}`);
  return response.json();
}

// =============================================================================
// REDDIT ENDPOINTS
// =============================================================================

/**
 * Get Reddit feed from crypto subreddits
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Reddit posts
 */
async function getRedditFeed(options = {}) {
  const params = new URLSearchParams({
    subreddit: options.subreddit || 'cryptocurrency',
    sort: options.sort || 'hot',
    limit: options.limit || 25,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/reddit?${params}`);
  return response.json();
}

/**
 * Get Reddit sentiment for an asset
 * @param {string} asset - Asset symbol
 * @returns {Promise<Object>} Reddit sentiment data
 */
async function getRedditSentiment(asset) {
  const response = await fetch(`${BASE_URL}/api/social/reddit/sentiment?asset=${asset}`);
  return response.json();
}

// =============================================================================
// YOUTUBE ENDPOINTS
// =============================================================================

/**
 * Get YouTube crypto videos
 * @param {Object} options - Query options
 * @returns {Promise<Object>} YouTube videos
 */
async function getYouTubeFeed(options = {}) {
  const params = new URLSearchParams({
    query: options.query || 'cryptocurrency news',
    limit: options.limit || 20,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/youtube?${params}`);
  return response.json();
}

// =============================================================================
// DISCORD ENDPOINTS
// =============================================================================

/**
 * Get Discord activity for crypto communities
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Discord activity
 */
async function getDiscordActivity(options = {}) {
  const params = new URLSearchParams({
    project: options.project || '',
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/discord?${params}`);
  return response.json();
}

// =============================================================================
// TELEGRAM ENDPOINTS
// =============================================================================

/**
 * Get Telegram channel feed
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Telegram messages
 */
async function getTelegramFeed(options = {}) {
  const params = new URLSearchParams({
    channel: options.channel || '',
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/telegram?${params}`);
  return response.json();
}

// =============================================================================
// GITHUB ENDPOINTS
// =============================================================================

/**
 * Get GitHub activity for crypto projects
 * @param {string} project - Project name or repo
 * @returns {Promise<Object>} GitHub activity
 */
async function getGitHubActivity(project) {
  const response = await fetch(`${BASE_URL}/api/social/github?project=${project}`);
  return response.json();
}

/**
 * Get developer activity metrics
 * @param {string} project - Project name
 * @returns {Promise<Object>} Developer metrics
 */
async function getDeveloperActivity(project) {
  const response = await fetch(`${BASE_URL}/api/social/github/developer-activity?project=${project}`);
  return response.json();
}

// =============================================================================
// INFLUENCERS & TRENDING
// =============================================================================

/**
 * Get top crypto influencers
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Influencer data
 */
async function getInfluencers(options = {}) {
  const params = new URLSearchParams({
    platform: options.platform || 'all',
    category: options.category || 'crypto',
    limit: options.limit || 50,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/influencers?${params}`);
  return response.json();
}

/**
 * Get aggregated social sentiment
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Social sentiment data
 */
async function getSocialSentiment(options = {}) {
  const params = new URLSearchParams({
    asset: options.asset || 'BTC',
    platforms: options.platforms?.join(',') || 'all',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/sentiment?${params}`);
  return response.json();
}

/**
 * Get trending topics across social media
 * @returns {Promise<Object>} Trending topics
 */
async function getSocialTrending() {
  const response = await fetch(`${BASE_URL}/api/social/trending`);
  return response.json();
}

/**
 * Get mentions volume for an asset
 * @param {string} asset - Asset symbol
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Mentions data
 */
async function getMentions(asset, options = {}) {
  const params = new URLSearchParams({
    asset,
    period: options.period || '24h',
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/mentions?${params}`);
  return response.json();
}

// =============================================================================
// GOVERNANCE
// =============================================================================

/**
 * Get governance proposals for a project
 * @param {string} project - Project name
 * @returns {Promise<Object>} Governance proposals
 */
async function getGovernance(project) {
  const response = await fetch(`${BASE_URL}/api/social/governance?project=${project}`);
  return response.json();
}

/**
 * Get specific governance proposal
 * @param {string} project - Project name
 * @param {string} proposalId - Proposal ID
 * @returns {Promise<Object>} Proposal details
 */
async function getProposal(project, proposalId) {
  const response = await fetch(
    `${BASE_URL}/api/social/governance/proposal?project=${project}&id=${proposalId}`
  );
  return response.json();
}

// =============================================================================
// EVENTS & CALENDAR
// =============================================================================

/**
 * Get upcoming crypto events
 * @param {Object} options - Query options
 * @returns {Promise<Object>} Events list
 */
async function getEvents(options = {}) {
  const params = new URLSearchParams({
    category: options.category || 'all',
    limit: options.limit || 20,
    ...options
  });
  const response = await fetch(`${BASE_URL}/api/social/events?${params}`);
  return response.json();
}

/**
 * Get crypto calendar for date range
 * @param {string} start - Start date (YYYY-MM-DD)
 * @param {string} end - End date (YYYY-MM-DD)
 * @returns {Promise<Object>} Calendar events
 */
async function getCalendar(start, end) {
  const response = await fetch(
    `${BASE_URL}/api/social/calendar?start=${start}&end=${end}`
  );
  return response.json();
}

// =============================================================================
// EXPORTS
// =============================================================================

module.exports = {
  // X/Twitter
  getXFeed,
  getXMentions,
  
  // Reddit
  getRedditFeed,
  getRedditSentiment,
  
  // YouTube
  getYouTubeFeed,
  
  // Discord
  getDiscordActivity,
  
  // Telegram
  getTelegramFeed,
  
  // GitHub
  getGitHubActivity,
  getDeveloperActivity,
  
  // Influencers & Trending
  getInfluencers,
  getSocialSentiment,
  getSocialTrending,
  getMentions,
  
  // Governance
  getGovernance,
  getProposal,
  
  // Events
  getEvents,
  getCalendar,
};

// =============================================================================
// EXAMPLE USAGE
// =============================================================================

async function runExamples() {
  console.log('\n=== Social Media API Examples ===\n');
  
  try {
    // X/Twitter feed
    console.log('1. Fetching X/Twitter feed...');
    const xFeed = await getXFeed({ topic: 'bitcoin', limit: 5 });
    console.log(`   Found ${xFeed.data?.length || 0} tweets`);
    
    // Reddit feed
    console.log('2. Fetching Reddit feed...');
    const reddit = await getRedditFeed({ subreddit: 'cryptocurrency', limit: 5 });
    console.log(`   Found ${reddit.data?.length || 0} posts`);
    
    // Social sentiment
    console.log('3. Getting social sentiment for BTC...');
    const sentiment = await getSocialSentiment({ asset: 'BTC' });
    console.log(`   Sentiment: ${sentiment.overall || 'N/A'}`);
    
    // Trending topics
    console.log('4. Fetching trending topics...');
    const trending = await getSocialTrending();
    console.log(`   Found ${trending.topics?.length || 0} trending topics`);
    
    // Influencers
    console.log('5. Fetching top influencers...');
    const influencers = await getInfluencers({ limit: 5 });
    console.log(`   Found ${influencers.data?.length || 0} influencers`);
    
    // GitHub activity
    console.log('6. Fetching GitHub activity for Bitcoin...');
    const github = await getGitHubActivity('bitcoin');
    console.log(`   Activity: ${github.commits || 0} recent commits`);
    
    // Governance
    console.log('7. Fetching governance proposals for Uniswap...');
    const governance = await getGovernance('uniswap');
    console.log(`   Found ${governance.proposals?.length || 0} proposals`);
    
    // Events
    console.log('8. Fetching upcoming events...');
    const events = await getEvents({ limit: 5 });
    console.log(`   Found ${events.data?.length || 0} events`);
    
    console.log('\n✅ All social examples completed!\n');
  } catch (error) {
    console.error('Error:', error.message);
  }
}

// Run if executed directly
if (require.main === module) {
  runExamples();
}

File: market-data/crypto-news/examples/python/social.py
#!/usr/bin/env python3
"""
Social & Community API Examples - Python
Free Crypto News API - https://github.com/nirholas/free-crypto-news

Examples for social media and community endpoints.
"""

import requests
import json
from typing import Optional, List

BASE_URL = "https://cryptocurrency.cv"


# =============================================================================
# GET /api/social/x - Twitter/X Feed
# =============================================================================

def get_x_feed(asset: Optional[str] = None, limit: int = 20) -> dict:
    """
    Get Twitter/X crypto posts.
    
    Args:
        asset: Filter by asset (BTC, ETH, etc.)
        limit: Number of posts
    
    Returns:
        X/Twitter posts
    """
    params = {"limit": limit}
    if asset:
        params["asset"] = asset
    
    response = requests.get(f"{BASE_URL}/api/social/x", params=params)
    return response.json()


# =============================================================================
# GET /api/social/reddit - Reddit Feed
# =============================================================================

def get_reddit_feed(subreddit: Optional[str] = None, limit: int = 20) -> dict:
    """
    Get Reddit crypto posts.
    
    Args:
        subreddit: Filter by subreddit (cryptocurrency, bitcoin, etc.)
        limit: Number of posts
    
    Returns:
        Reddit posts
    """
    params = {"limit": limit}
    if subreddit:
        params["subreddit"] = subreddit
    
    response = requests.get(f"{BASE_URL}/api/social/reddit", params=params)
    return response.json()


# =============================================================================
# GET /api/social/youtube - YouTube Feed
# =============================================================================

def get_youtube_feed(channel: Optional[str] = None, limit: int = 20) -> dict:
    """
    Get YouTube crypto videos.
    
    Args:
        channel: Filter by channel
        limit: Number of videos
    
    Returns:
        YouTube videos
    """
    params = {"limit": limit}
    if channel:
        params["channel"] = channel
    
    response = requests.get(f"{BASE_URL}/api/social/youtube", params=params)
    return response.json()


# =============================================================================
# GET /api/social/discord - Discord Activity
# =============================================================================

def get_discord_activity() -> dict:
    """
    Get Discord community activity.
    
    Returns:
        Discord activity metrics
    """
    response = requests.get(f"{BASE_URL}/api/social/discord")
    return response.json()


# =============================================================================
# GET /api/social/telegram - Telegram Feed
# =============================================================================

def get_telegram_feed(channel: Optional[str] = None, limit: int = 20) -> dict:
    """
    Get Telegram crypto channel posts.
    
    Args:
        channel: Filter by channel
        limit: Number of posts
    
    Returns:
        Telegram posts
    """
    params = {"limit": limit}
    if channel:
        params["channel"] = channel
    
    response = requests.get(f"{BASE_URL}/api/social/telegram", params=params)
    return response.json()


# =============================================================================
# GET /api/social/github - GitHub Activity
# =============================================================================

def get_github_activity(project: Optional[str] = None) -> dict:
    """
    Get GitHub crypto project activity.
    
    Args:
        project: Filter by project name
    
    Returns:
        GitHub activity metrics
    """
    params = {}
    if project:
        params["project"] = project
    
    response = requests.get(f"{BASE_URL}/api/social/github", params=params)
    return response.json()


# =============================================================================
# GET /api/social/developers - Developer Activity
# =============================================================================

def get_developer_activity(asset: str = "ethereum") -> dict:
    """
    Get developer activity for a project.
    
    Args:
        asset: Asset/project name
    
    Returns:
        Developer activity metrics
    """
    response = requests.get(f"{BASE_URL}/api/social/developers", 
                           params={"asset": asset})
    return response.json()


# =============================================================================
# GET /api/social/influencers - Crypto Influencers
# =============================================================================

def get_influencers(limit: int = 20) -> dict:
    """
    Get list of crypto influencers.
    
    Args:
        limit: Number of influencers
    
    Returns:
        Influencer list with metrics
    """
    response = requests.get(f"{BASE_URL}/api/social/influencers", 
                           params={"limit": limit})
    return response.json()


# =============================================================================
# GET /api/social/sentiment - Social Sentiment
# =============================================================================

def get_social_sentiment(asset: str = "BTC") -> dict:
    """
    Get social media sentiment analysis.
    
    Args:
        asset: Asset symbol
    
    Returns:
        Social sentiment metrics
    """
    response = requests.get(f"{BASE_URL}/api/social/sentiment", 
                           params={"asset": asset})
    return response.json()


# =============================================================================
# GET /api/social/trending - Trending on Social
# =============================================================================

def get_social_trending() -> dict:
    """
    Get trending topics across social platforms.
    
    Returns:
        Trending topics
    """
    response = requests.get(f"{BASE_URL}/api/social/trending")
    return response.json()


# =============================================================================
# GET /api/social/mentions - Mention Tracking
# =============================================================================

def get_mentions(asset: str = "BTC", period: str = "24h") -> dict:
    """
    Get mention tracking for an asset.
    
    Args:
        asset: Asset symbol
        period: Time period
    
    Returns:
        Mention counts and sources
    """
    params = {"asset": asset, "period": period}
    response = requests.get(f"{BASE_URL}/api/social/mentions", params=params)
    return response.json()


# =============================================================================
# GET /api/governance - Governance Updates
# =============================================================================

def get_governance(protocol: Optional[str] = None, limit: int = 20) -> dict:
    """
    Get DAO governance proposals and votes.
    
    Args:
        protocol: Filter by protocol (aave, compound, etc.)
        limit: Number of proposals
    
    Returns:
        Governance proposals
    """
    params = {"limit": limit}
    if protocol:
        params["protocol"] = protocol
    
    response = requests.get(f"{BASE_URL}/api/governance", params=params)
    return response.json()


def get_proposal(proposal_id: str) -> dict:
    """
    Get details of a specific proposal.
    
    Args:
        proposal_id: Proposal ID
    
    Returns:
        Proposal details
    """
    response = requests.get(f"{BASE_URL}/api/governance/{proposal_id}")
    return response.json()


# =============================================================================
# GET /api/events - Crypto Events
# =============================================================================

def get_events(type: Optional[str] = None, limit: int = 20) -> dict:
    """
    Get upcoming crypto events.
    
    Args:
        type: Event type (conference, ama, airdrop, etc.)
        limit: Number of events
    
    Returns:
        Upcoming events
    """
    params = {"limit": limit}
    if type:
        params["type"] = type
    
    response = requests.get(f"{BASE_URL}/api/events", params=params)
    return response.json()


# =============================================================================
# GET /api/calendar - Events Calendar
# =============================================================================

def get_calendar(month: Optional[int] = None, year: Optional[int] = None) -> dict:
    """
    Get crypto events calendar.
    
    Args:
        month: Month number (1-12)
        year: Year
    
    Returns:
        Calendar events
    """
    params = {}
    if month:
        params["month"] = month
    if year:
        params["year"] = year
    
    response = requests.get(f"{BASE_URL}/api/calendar", params=params)
    return response.json()


# =============================================================================
# COMPLETE EXAMPLES
# =============================================================================

if __name__ == "__main__":
    print("\n" + "="*60)
    print("FREE CRYPTO NEWS API - SOCIAL & COMMUNITY EXAMPLES")
    print("="*60)
    
    # 1. X/Twitter
    print("\n🐦 1. X/Twitter Feed")
    x_feed = get_x_feed(limit=5)
    print(f"   X Posts: {x_feed}")
    
    # 2. Reddit
    print("\n🔴 2. Reddit Feed")
    reddit = get_reddit_feed(subreddit="cryptocurrency", limit=5)
    print(f"   Reddit: {reddit}")
    
    # 3. YouTube
    print("\n📺 3. YouTube Videos")
    yt = get_youtube_feed(limit=5)
    print(f"   YouTube: {yt}")
    
    # 4. Telegram
    print("\n📱 4. Telegram Feed")
    tg = get_telegram_feed(limit=5)
    print(f"   Telegram: {tg}")
    
    # 5. GitHub
    print("\n💻 5. GitHub Activity")
    gh = get_github_activity(project="ethereum")
    print(f"   GitHub: {gh}")
    
    # 6. Influencers
    print("\n👤 6. Top Influencers")
    influencers = get_influencers(limit=5)
    print(f"   Influencers: {influencers}")
    
    # 7. Social Sentiment
    print("\n😊 7. Social Sentiment (BTC)")
    sentiment = get_social_sentiment("BTC")
    print(f"   Sentiment: {sentiment}")
    
    # 8. Trending
    print("\n🔥 8. Social Trending")
    trending = get_social_trending()
    print(f"   Trending: {trending}")
    
    # 9. Mentions
    print("\n📢 9. BTC Mentions (24h)")
    mentions = get_mentions("BTC", "24h")
    print(f"   Mentions: {mentions}")
    
    # 10. Governance
    print("\n🏛️ 10. DAO Governance")
    gov = get_governance(limit=5)
    print(f"   Governance: {gov}")
    
    # 11. Events
    print("\n📅 11. Upcoming Events")
    events = get_events(limit=5)
    print(f"   Events: {events}")
    
    print("\n" + "="*60)
    print("All social & community examples completed!")
    print("="*60)
